scipy/sparse/construct.py

```diff
@@ -644,38 +645,73 @@ def block_diag(mats, format=None, dtype=None):
 
     Notes
     -----
+    Providing a sequence of equally shaped matrices
+     will provide marginally faster results
 
-    .. versionadded:: 0.11.0
+    .. versionadded:: 0.18.0
 
     See Also
     --------
-    bmat, diags
+    bmat, diags, block_diag
 
     Examples
     --------
     >>> from scipy.sparse import coo_matrix, block_diag
     >>> A = coo_matrix([[1, 2], [3, 4]])
-    >>> B = coo_matrix([[5], [6]])
-    >>> C = coo_matrix([[7]])
+    >>> B = coo_matrix([[5, 6], [7, 8]])
+    >>> C = coo_matrix([[9, 10], [11,12]])
     >>> block_diag((A, B, C)).toarray()
-    array([[1, 2, 0, 0],
-           [3, 4, 0, 0],
-           [0, 0, 5, 0],
-           [0, 0, 6, 0],
-           [0, 0, 0, 7]])
-
+    array([[ 1,  2,  0,  0,  0,  0],
+           [ 3,  4,  0,  0,  0,  0],
+           [ 0,  0,  5,  6,  0,  0],
+           [ 0,  0,  7,  8,  0,  0],
+           [ 0,  0,  0,  0,  9, 10],
+           [ 0,  0,  0,  0, 11, 12]])
     """
-    nmat = len(mats)
-    rows = []
+
+    n = len(mats)
+    mats_ = [None] * n
     for ia, a in enumerate(mats):
-        row = [None]*nmat
-        if issparse(a):
-            row[ia] = a
+        if hasattr(a, 'shape'):
+            mats_[ia] = a
         else:
-            row[ia] = coo_matrix(a)
-        rows.append(row)
-    return bmat(rows, format=format, dtype=dtype)
-
+            mats_[ia] = coo_matrix(a)
+
+    if any(mat.shape != mats_[-1].shape for mat in mats_) or (
+            any(issparse(mat) for mat in mats_)):
+        data = []
+        col = []
+        row = []
+        origin = np.array([0, 0], dtype=np.int)
+        for mat in mats_:
+            if issparse(mat):
+                data.append(mat.data)
+                row.append(mat.row + origin[0])
+                col.append(mat.col + origin[1])
+
+            else:
+                data.append(mat.ravel())
```

Comment:
I suggest writing `mat = np.ravel(mat)` --- the input might not actually be a numpy array, so the existence of `.ravel()` and that it works correctly is not given.